eth/db/account.py

```diff
@@ -390,30 +356,42 @@ def _set_account(self, address: Address, account: Account) -> None:
     #
     # Record and discard API
     #
-    def record(self) -> Tuple[UUID, UUID]:
-        return (self._journaldb.record(), self._journaltrie.record())
-
-    def discard(self, changeset: Tuple[UUID, UUID]) -> None:
-        db_changeset, trie_changeset = changeset
-        self._journaldb.discard(db_changeset)
-        self._journaltrie.discard(trie_changeset)
+    def record(self) -> UUID:
+        changeset_id = self._journaldb.record()
+        self._journaltrie.record(changeset_id)
+        return changeset_id
+
+    def discard(self, changeset: UUID) -> None:
+        self._journaldb.discard(changeset)
+        self._journaltrie.discard(changeset)
         self._account_cache.clear()
 
-    def commit(self, changeset: Tuple[UUID, UUID]) -> None:
-        db_changeset, trie_changeset = changeset
-        self._journaldb.commit(db_changeset)
-        self._journaltrie.commit(trie_changeset)
+    def commit(self, changeset: UUID) -> None:
+        self._journaldb.commit(changeset)
+        self._journaltrie.commit(changeset)
 
     def make_state_root(self) -> Hash32:
-        self.logger.debug2("Generating AccountDB trie")
         self._journaldb.persist()
         self._journaltrie.persist()
         return self.state_root
 
     def persist(self) -> None:
-        self.make_state_root()
-        self._batchtrie.commit(apply_deletes=False)
-        self._batchdb.commit(apply_deletes=True)
+        self._validate_generated_root()
+        with self._raw_store_db.atomic_batch() as write_batch:
+            self._batchtrie.commit_to(write_batch, apply_deletes=False)
```

Comment:
Won't `apply_deletes=False` orphan a lot of intermediate nodes?

Reply:
Yes, and yes. But we don't know which recent tries still need access to those intermediate nodes. Also, we don't know what other accounts are still using the code with that hash. For now this is the easy/safe approach.

Reply:
Ahh, really good point re multiple contracts might be sharing that code!